tests/openedx_tagging/core/tagging/test_api.py

```diff
@@ -0,0 +1,218 @@
+        self.taxonomy.allow_multiple = True
+        test_tags = [
+            [
+                self.archaea.id,
+                self.eubacteria.id,
+                self.chordata.id,
+            ],
+            [
+                self.chordata.id,
+                self.archaebacteria.id,
+            ],
+            [
+                self.archaebacteria.id,
+                self.archaea.id,
+            ],
+        ]
+
+        # Tag and re-tag the object, checking that the expected tags are returned and deleted
+        for tag_list in test_tags:
+            object_tags = tagging_api.tag_object(
+                self.taxonomy,
+                tag_list,
+                "biology101",
+                "course",
+            )
+
+            # Ensure the expected number of tags exist in the database
+            assert tagging_api.get_object_tags(
+                taxonomy=self.taxonomy,
+                object_id="biology101",
+                object_type="course",
+            ) == object_tags
+            # And the expected number of tags were returned
+            assert len(object_tags) == len(tag_list)
+            for index, object_tag in enumerate(object_tags):
+                assert object_tag.tag_id == tag_list[index]
+                assert object_tag.is_valid
+                assert object_tag.taxonomy == self.taxonomy
+                assert object_tag.name == self.taxonomy.name
+                assert object_tag.object_id == "biology101"
+                assert object_tag.object_type == "course"
+
+    def test_autocomplete_tags_closed_taxonomy(self):
+        object_id = 'course_id_1'
+        prefix = 'a'
+        self._validate_autocomplete_tags(self.taxonomy, prefix, object_id)
+
+    def test_autocomplete_tags_free_text_taxonomy(self):
+        object_id_1 = 'course_id_1'
+        object_id_2 = 'course_id_2'
+        prefix = 'a'
+
+        taxonomy = tagging_api.create_taxonomy("Free_Text_Taxonomy", allow_free_text=True)
+        tags = [
+            'Archaea',
+            'Archaebacteria',
+            'Animalia',
+            'Arthropoda',
+            'Plantae',
+            'Monera',
+            'Gastrotrich',
+            'Placozoa',
+            # For testing repeats
+            'Animalia',
+            'Arthropoda',
+        ]
+        for tag in tags:
+            ObjectTag(
+                object_id=object_id_1,
+                object_type='course',
+                taxonomy=taxonomy,
+                _value=tag,
+            ).save()
+
+        self._validate_autocomplete_tags(taxonomy, prefix, object_id_2, free_text=True)
+
+    def _validate_autocomplete_tags(self, taxonomy, prefix, object_id, free_text=False):
+        """
+        Validate autocomplete tags
+        """
+
+        # Normal search
+        result = tagging_api.autocomplete_tags(taxonomy, prefix)
+        self.assertEqual(len(result), 4)
+        for tag in result:
+            self.assertEqual(tag[0].lower(), prefix)
```

Comment:
Can you check that the tags are returned in the expected order?

```suggestion
        self.assertEqual(result, ["Animalia", "Archaea", "Archaebacteria", "Arthropoda"])
```